libs/posthog_client_lite/src/lib.rs

```diff
@@ -0,0 +1,433 @@
+                    return None;
+                }
+            }
+            // If all conditions are met, return the variant
+            if self.evaluate_percentage(mapped_user_id, group.rollout_percentage / 100.0) {
+                group.variant.clone()
+            } else {
+                None
+            }
+        } else {
+            // No matchers, apply to all users
+            if self.evaluate_percentage(mapped_user_id, group.rollout_percentage / 100.0) {
+                group.variant.clone()
+            } else {
+                None
+            }
+        }
+    }
+
+    /// Evaluate a multivariate feature flag. Returns `None` if the flag is not available or if there are errors
+    /// during the evaluation.
+    ///
+    /// The parsing logic is as follows:
+    ///
+    /// * Match each filter group. If any group is matched, check the rollout percentage against the precomputed
+    ///   user ID on the consistent hash ring. If the user ID is in the range of the group's rollout percentage,
+    ///   return the variant.
+    /// * Otherwise, continue with the next group until all groups are evaluated and no group is within the
+    ///   rollout percentage.
+    /// * If there are no matching groups, evaluate by the global rollout percentage.
+    pub fn evaluate_multivariate(&self, flag_key: &str, user_id: &str) -> Option<String> {
+        let mapped_user_id = Self::consistent_hash(user_id);
+        self.evaluate_multivariate_inner(flag_key, mapped_user_id, &HashMap::new())
+    }
+
+    /// Evaluate a multivariate feature flag. Returns `None` if the flag is not available or if there are errors
+    /// during the evaluation. Note that we directly take the mapped user ID (a consistent hash ranging from 0 to 1)
+    /// so that it is easier to use it in the tests and avoid duplicate computations.
+    pub fn evaluate_multivariate_inner(
+        &self,
+        flag_key: &str,
+        mapped_user_id: f64,
+        properties: &HashMap<String, PostHogFlagFilterPropertyValue>,
+    ) -> Option<String> {
+        if let Some(flag_config) = self.flags.get(flag_key) {
+            // First, try to evaluate each group
+            for group in &flag_config.filters.groups {
+                if let Some(variant) = self.evaluate_group(group, mapped_user_id, properties) {
+                    return Some(variant);
+                }
+            }
+            // If no group is matched, evaluate by the global rollout percentage
+            let mut percentage = 0.0;
+            for variant in &flag_config.filters.multivariate.variants {
+                percentage += variant.rollout_percentage;
+                if self.evaluate_percentage(mapped_user_id, percentage / 100.0) {
+                    return Some(variant.key.clone());
+                }
+            }
+            // This should not happen because the rollout percentage always adds up to 100, but just in case,
```

Comment:
Well, I guess it's very unlikely to happen, but working with floating numbers, I wouldn't assume that some numbers add up to 100. There is always a precision error. Maybe returning the last value would be a better option?

Reply:
PostHog always gives integer as percentage so actually we can also sum using i64 here. I will switch to i64 instead to ensure there're no precision issues.